### PR TITLE
#170: Rewrite ImportService

### PR DESCRIPTION
**Pull request by kcissewski**

### CODE REVIEW

File: src/main/java/org/benetech/servicenet/service/impl/TmpImportServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,651 @@
...
+        createOrUpdateProgramsForOrganization(filledOrganization.getPrograms(), organization);
+
+        registerSynchronizationOfMatchingOrganizations(organization);
+
+        return organization;
+    }
+
+    @Override
+    @ConfidentialFilter
+    @Transactional(propagation = Propagation.REQUIRED)
```
Comment (oskarhinc) [inline src/main/java/org/benetech/servicenet/service/impl/TmpImportServiceImpl.java]: Are you sure the propagation should be REQUIRED?

File: src/main/java/org/benetech/servicenet/service/impl/TmpImportServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+    private void registerSynchronizationOfMatchingOrganizations(Organization organization) {
+        TransactionSynchronizationManager.registerSynchronization(new TransactionSynchronizationAdapter() {
+            @Override
+            public void afterCommit() {
+                organizationMatchService.createOrUpdateOrganizationMatches(organization);
+            }
+        });
+    }
+
+    private void persistWithoutUnsavedRelatedEntity(Phone phone) {
```
Comment (oskarhinc) [inline src/main/java/org/benetech/servicenet/service/impl/TmpImportServiceImpl.java]: What's the purpose of those methods?

File: src/main/java/org/benetech/servicenet/service/mapper/LocationMapper.java [hunk 3]
```diff
@@ -23,6 +23,10 @@
     @Mapping(target = "accessibilities", ignore = true)
     Location toEntity(LocationDTO locationDTO);
 
+    @Mapping(source = "organization.id", target = "organizationId")
+    @Mapping(source = "organization.name", target = "organizationName")
+    LocationDTO toDto(Location service);
```
Comment (oskarhinc) [inline src/main/java/org/benetech/servicenet/service/mapper/LocationMapper.java]: Location service?

File: src/main/resources/.h2.server.properties [hunk 4]
```diff
@@ -1,5 +1,5 @@
 #H2 Server Properties
-#Fri Jan 11 14:42:47 CET 2019
+#Wed Jan 30 11:09:34 CET 2019
```
Comment (oskarhinc) [inline src/main/resources/.h2.server.properties]: unwanted change

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 5]
```diff
@@ -112,153 +172,220 @@ public void shouldUpdateLocation() {
...
-        int dbSize = locationService.findAll().size();
-        Location updated = importService.createOrUpdateLocation(newLocation, EXISTING_EXTERNAL_ID, PROVIDER);
+        assertEquals(1, locationService.findAll().size());
+        importService.createOrUpdateLocation(newLocation, EXISTING_EXTERNAL_ID, PROVIDER);
 
-        assertEquals(dbSize, locationService.findAll().size());
-        assertEquals(NEW_STRING, updated.getName());
+        List<LocationDTO> all = locationService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(l -> l.getName().equals(NEW_STRING)));
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: 'l' is not the best name for a variable

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 6]
```diff
@@ -112,153 +172,220 @@ public void shouldUpdateLocation() {
...
 
-        int dbSize = physicalAddressService.findAll().size();
-        importService.createOrUpdatePhysicalAddress(address, location);
+        assertEquals(0, physicalAddressService.findAll().size());
+        importService.createOrUpdateLocation(location, EXISTING_EXTERNAL_ID, PROVIDER);
 
-        assertEquals(dbSize + 1, physicalAddressService.findAll().size());
+        List<PhysicalAddressDTO> all = physicalAddressService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(pa -> pa.getLocationId() != null
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: I think it's better to extract the var and assert all individually, to point out specifically what goes wrong if the test won't pass in the future.

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 7]
```diff
@@ -112,153 +172,220 @@ public void shouldUpdateLocation() {
...
+            .attention(NEW_STRING).country(NEW_STRING).postalCode(NEW_STRING).region(NEW_STRING);
+        locationToUpdate.setPhysicalAddress(newAddress);
+
+        assertEquals(1, locationService.findAll().size());
+
+        importService.createOrUpdateLocation(locationToUpdate, EXISTING_EXTERNAL_ID, PROVIDER);
+
+        List<PhysicalAddressDTO> all = physicalAddressService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(pa -> pa.getLocationId() != null
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: same here

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 8]
```diff
@@ -112,153 +172,220 @@ public void shouldUpdateLocation() {
...
+
+        assertEquals(0, postalAddressService.findAll().size());
+
+        importService.createOrUpdateLocation(location, NEW_EXTERNAL_ID, PROVIDER);
+
+        assertEquals(1, locationService.findAll().size());
+        List<PostalAddressDTO> all = postalAddressService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(pa -> pa.getLocationId() != null
+            && pa.getAddress1().equals(NEW_STRING)
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: same

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 9]
```diff
@@ -112,153 +172,220 @@ public void shouldUpdateLocation() {
...
+        locationToUpdate.setPostalAddress(newAddress);
+
+        assertEquals(1, locationService.findAll().size());
+
+        importService.createOrUpdateLocation(locationToUpdate, EXISTING_EXTERNAL_ID, PROVIDER);
+
+        assertEquals(1, locationService.findAll().size());
+        List<PostalAddressDTO> all = postalAddressService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(pa -> pa.getLocationId() != null
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: same

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 10]
```diff
@@ -323,47 +464,125 @@ public void shouldReplacePhonesIfServiceHasFewOfThemButNotThisOne() {
...
+        Service serviceToUpdate = generateExistingServiceDoNotPersist();
+        serviceToUpdate.setPhones(mutableSet(newPhone));
+
+        assertEquals(2, phoneService.findAll().size());
+
+        importService.createOrUpdateService(serviceToUpdate, EXISTING_EXTERNAL_ID, PROVIDER, new DataImportReport());
+
+        List<PhoneDTO> all = phoneService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(phone -> phone.getNumber().equals(NEW_STRING)
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: same

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 11]
```diff
@@ -323,47 +464,125 @@ public void shouldReplacePhonesIfServiceHasFewOfThemButNotThisOne() {
...
+        Phone phone = new Phone().number(NEW_STRING).location(location);
+        location.setPhones(mutableSet(phone));
+
+        assertEquals(0, phoneService.findAll().size());
+
+        importService.createOrUpdateLocation(location, EXISTING_EXTERNAL_ID, PROVIDER);
+
+        List<PhoneDTO> all = phoneService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(p -> p.getNumber().equals(NEW_STRING)
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: same

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 12]
```diff
@@ -323,47 +464,125 @@ public void shouldReplacePhonesIfServiceHasFewOfThemButNotThisOne() {
...
+        assertEquals(1, phoneService.findAll().size());
 
-        assertEquals(dbSize - 1, phoneService.findAll().size());
-        assertEquals(1, updated.size());
-        assertTrue(updated.stream().anyMatch(phone -> phone.equals(newPhone)));
+        importService.createOrUpdateLocation(locationToUpdate, EXISTING_EXTERNAL_ID, PROVIDER);
+
+        List<PhoneDTO> all = phoneService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(p -> p.getNumber().equals(NEW_STRING)
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: same

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 13]
```diff
@@ -323,47 +464,125 @@ public void shouldReplacePhonesIfServiceHasFewOfThemButNotThisOne() {
...
+        Location locationToUpdate = generateExistingLocationDoNotPersist();
+        locationToUpdate.setPhones(mutableSet(newPhone));
+
+        assertEquals(2, phoneService.findAll().size());
+
+        importService.createOrUpdateLocation(locationToUpdate, EXISTING_EXTERNAL_ID, PROVIDER);
+
+        List<PhoneDTO> all = phoneService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(phone -> phone.getNumber().equals(NEW_STRING)
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: same

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 14]
```diff
@@ -323,47 +464,125 @@ public void shouldReplacePhonesIfServiceHasFewOfThemButNotThisOne() {
...
-        int dbSize = eligibilityService.findAll().size();
-        importService.createOrUpdateEligibility(eligibility, service);
+        assertEquals(0, eligibilityService.findAll().size());
 
-        assertEquals(dbSize + 1, eligibilityService.findAll().size());
+        importService.createOrUpdateService(service, EXISTING_EXTERNAL_ID, PROVIDER, new DataImportReport());
+
+        List<EligibilityDTO> all = eligibilityService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(e -> e.getEligibility().equals(NEW_STRING)
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: same

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 15]
```diff
@@ -375,26 +594,38 @@ public void shouldUpdateEligibility() {
...
+            .services(Collections.singleton(service));
 
-        assertEquals(dbSize + 1, languageService.findAll().size());
+        assertEquals(0, languageService.findAll().size());
+
+        importService.createOrUpdateOrganization(org, EXISTING_EXTERNAL_ID, PROVIDER, new DataImportReport());
+
+        List<LanguageDTO> all = languageService.findAll();
+        assertEquals(1, all.size());
+        assertTrue(all.stream().allMatch(l -> l.getLanguage().equals(NEW_STRING)
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: see other occurrences

File: src/main/java/org/benetech/servicenet/service/impl/TmpImportServiceImpl.java (hunk 2 above)
Comment (kcissewski) [reply src/main/java/org/benetech/servicenet/service/impl/TmpImportServiceImpl.java]: Discussed offline.

File: src/main/java/org/benetech/servicenet/service/impl/TmpImportServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,651 @@
...
+
+import javax.annotation.Nonnull;
+import javax.persistence.EntityManager;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Component
+public class TmpImportServiceImpl implements TmpImportService {
```
[the remark this answers is not in the document]
Comment (kcissewski) [reply src/main/java/org/benetech/servicenet/service/impl/TmpImportServiceImpl.java]: Discussed offline

**Comment by kcissewski** [comment]
@oskarhinc PR updated, please review when you will have time. Thanks!

File: src/main/java/org/benetech/servicenet/service/SharedLogicService.java [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+
+import org.benetech.servicenet.domain.Location;
+import org.benetech.servicenet.domain.OpeningHours;
+import org.benetech.servicenet.domain.Phone;
+import org.benetech.servicenet.domain.RegularSchedule;
+import org.benetech.servicenet.domain.Service;
+
+import java.util.Set;
+
+public interface SharedLogicService {
```
Comment (oskarhinc) [inline src/main/java/org/benetech/servicenet/service/SharedLogicService.java]: Maybe SharedImportService? As it's in the package where import-related classes are not the only services?

File: src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java [hunk 18]
```diff
@@ -533,10 +1361,44 @@ private Organization generateNewOrganization(SystemAccount account) {
...
+        Taxonomy result = new Taxonomy().externalDbId(EXISTING_EXTERNAL_ID).providerName(PROVIDER)
+            .name(OTHER_STRING);
+        em.persist(result);
+        return result;
+    }
+
+    @SafeVarargs
+    private <T> Set<T> mutableSet(T ... objects) {
+        return new HashSet<T>(Arrays.asList(objects));
+    }
```
Comment (oskarhinc) [inline src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: I'm afraid this class should be split into smaller ones, just like normal services :face_with_head_bandage:

Comment (oskarhinc) [reply src/test/java/org/benetech/servicenet/service/impl/ImportServiceImplIntTest.java]: But we can definitely do this in a separate PR, but we have to remember it.

**Comment by oskarhinc** [comment]
As the suggestions do not affect the business logic, and the PR gets really big, I'll merge this one, but the changes should be applied in a separate one.

**Pull request merged**